### PR TITLE
Report permissions denied in internal SSH

### PR DESCRIPTION
**Pull request by zeripath**
This PR standardizes reporting of permission denied from the internal ssh.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by mrsdizzie — APPROVED** [review]

**Review by adelowo — APPROVED** [review]

**Comment by zeripath** [comment]
I've marked this as backport as I think this is likely to be helpful to other places

File: modules/ssh/ssh.go [hunk 1]
```diff
@@ -170,11 +172,14 @@ func publicKeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
 
 			// check the CA of the cert
 			if !c.IsUserAuthority(cert.SignatureKey) {
-				return false
+				log.Debug("Principal Rejected: Untrusted Authority Signature Fingerprint %s for Principal: %s", gossh.FingerprintSHA256(cert.SignatureKey), principal)
```
Comment (zeripath) [inline modules/ssh/ssh.go:175]: So this should really be wrapped in an ```
if log.IsDebug() { ... }
```

File: modules/ssh/ssh.go [hunk 2]
```diff
@@ -170,11 +172,14 @@ func publicKeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
...
+				continue principalLoop
 			}
 
 			// validate the cert for this principal
 			if err := c.CheckCert(principal, cert); err != nil {
+				// User is presenting an invalid cerficate - STOP any further processing
+				log.Error("Permission Denied: Invalid Certificate KeyID %s with Signature Fingerprint %s presented for Principal: %s", cert.KeyId, gossh.FingerprintSHA256(cert.SignatureKey), principal)
```
Comment (zeripath) [inline modules/ssh/ssh.go:182]: As error is so commonly set we can probably get away without a log.IsError() around this

File: modules/ssh/ssh.go [hunk 3]
```diff
@@ -186,6 +191,10 @@ func publicKeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
 
 	pkey, err := models.SearchPublicKeyByContent(strings.TrimSpace(string(gossh.MarshalAuthorizedKey(key))))
 	if err != nil {
+		if models.IsErrKeyNotExist(err) {
+			log.Warn("Permission Denied: Unknown public key : %s", gossh.FingerprintSHA256(key))
```
Comment (zeripath) [inline modules/ssh/ssh.go:195]: Similar to the debug case

File: modules/ssh/ssh.go [hunk 4]
```diff
@@ -145,17 +146,18 @@ func publicKeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
 		}
 
 		// look for the exact principal
+	principalLoop:
 		for _, principal := range cert.ValidPrincipals {
 			pkey, err := models.SearchPublicKeyByContentExact(principal)
 			if err != nil {
+				if models.IsErrKeyNotExist(err) {
+					log.Debug("Principal Rejected: Unknown Principal: %s", principal)
+					continue principalLoop
```
Comment (lunny) [inline modules/ssh/ssh.go:155]: Isn't this will loop forever?

Comment (zeripath) [reply modules/ssh/ssh.go:155]: No.

https://play.golang.org/p/HA0jVr08_TV

**Comment by zeripath** [comment]
I've removed backport as it would break fail2ban configurations in the wild.

**Pull request merged**